File: Level_4(C++,Java)/C++/Gun/Gun.cpp
#include "Gun.h"

Gun::Gun(const std::string& model, int capacity) {
    this->amount = 0;
    this->capacity = capacity;
    this->isReady = 0;
    this->model = model;
    this->totalShots = 0;
}

Gun::~Gun() {}

int Gun::getAmount() const {
    return this->amount;
}

int Gun::getCapacity() const {
    return this->capacity;
}

bool Gun::ready() const {
    return this->isReady;
}

const std::string& Gun::getModel() const {
    return this->model;
}

int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    this->isReady = !(this->isReady);
    
    std::cout << this->model << " prepare!" << std::endl;
}

void Gun::reload() {
    this->amount = this->capacity;
    this->isReady = 0;
    
    std::cout << this->model << " reloded!" << std::endl;
}

void Gun::shoot() {
    if ( this->amount == 0 ) {
        throw OutOfRounds();
    }
    if ( !(this->isReady) ) {
        throw NotReady();
    }
    
    this->totalShots += 1;
    this->amount -= 1;
    
    std::cout << "Bang!" << std::endl;
}

std::ostream& operator<<(std::ostream& out, const Gun& gun) {
    out << gun.getModel() << ": " << "Ammunition[" << gun.getAmount() << '/' << gun.getCapacity() << "] Ready[";
    
    if ( gun.ready() ) {
        out << "yes] ";
    } else {
        out << "no] ";
    }
    
    out << "Total shots[" << gun.getTotalShots() << ']';
}

File: Level_4(C++,Java)/java/Gun/Gun.java
public class Gun {
    private int amount;
    private int capacity;
    private String model;
    private boolean isReady;
    private int totalShots;
    
    public Gun(String model, int capacity) {
        this.amount = 0;
        this.capacity = capacity;
        this.model = model;
        this.isReady = false;
        this.totalShots = 0;
    }
    
    public Gun() {
        this("Beretta", 8);
    }
    
    public int getAmount() {
        return this.amount;
    }
    
    public int getCapacity() {
        return this.capacity;
    }
    
    public boolean ready() {
        return this.isReady;
    }
    
    public String getModel() {
        return this.model;
    }
    
    public int getTotalShots() {
        return this.totalShots;
    }
    
    public void prepare() {
        this.isReady = !(this.isReady);
    
        System.out.println(this.model + " prepare!");
    }
    
    public void reload() {
        this.amount = this.capacity;
        this.isReady = false;
        
        System.out.println(this.model + " reloded!");
    }
    
    public void shoot() throws OutOfRoundsException, NotReadyException {
        if ( this.amount == 0 ) {
            throw new OutOfRoundsException();
        }
        if ( !(this.isReady) ) {
            throw new NotReadyException();
        }
        
        this.totalShots += 1;
        this.amount -= 1;
        
        System.out.println("Bang!");
    }
    
    public String toString() {
        StringBuffer out = new StringBuffer();
        
        out.append(this.model);
        out.append(": ");
        out.append("Ammunition[");
        out.append(this.amount);
        out.append('/');
        out.append(this.capacity);
        out.append("] Ready[");
        
        if ( this.isReady ) {
            out.append("yes] ");
        } else {
            out.append("no] ");
        }
        
        out.append("Total shots[");
        out.append(this.totalShots);
        out.append(']');
        
        return out.toString();
    }
}
